spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
   but may also depend on other values or operations.
 * *w<sup>r</sup><sub>p</sub> = w<sup>r</sup><sub>q</sub> &#8655;
   e<sup>r</sup><sub>p</sub> = e<sup>r</sup><sub>q</sub>*
+
+## Managing the Application state and related topics
+
+### Connection State
+
+Tendermint maintains four concurrent ABCI++ connections, namely
+[Consensus Connection](#consensus-connection),
+[Mempool Connection](#mempool-connection),
+[Info/Query Connection](#infoquery-connection), and
+[Snapshot Connection](#snapshot-connection).
+It is common for an application to maintain a distinct copy of
+the state for each connection, which are synchronized upon `Commit` calls.
+
+#### Concurrency
+
+In principle, each of the four ABCI++ connections operates concurrently with one
+another. This means applications need to ensure access to state is
+thread safe. Up to v0.35.x, both the
+[default in-process ABCI client](https://github.com/tendermint/tendermint/blob/v0.35.x/abci/client/local_client.go#L18)
+and the
+[default Go ABCI server](https://github.com/tendermint/tendermint/blob/v0.35.x/abci/server/socket_server.go#L32)
+used global locks to guard the handling of events across all connections, so they were not
+concurrent at all. This meant if your app was written in Go, and compiled in-process with Tendermint
+using the default `NewLocalClient`, or run out-of-process using the default `SocketServer`,
+ABCI messages from all connections were linearizable (received one at a
+time).
+This is no longer the case starting from v0.36.1: the global locks have been removed and it is
+up to the Application to ensure that handling ABCI++ connections is thread safe.
+
+#### FinalizeBlock
+
+When the Consensus algorithm decides on a block, Tendermint uses `FinalizeBlock` to send the
+current block's hash, height, transactions, etc., which the Application uses to transition its state.
+
+The Application should remember the latest height from which it
+has run a successful `Commit` so that it can tell Tendermint where to
+pick up from when it restarts. See information on the Handshake, below.
+
+#### Commit
+
+The Application state should be persisted during `Commit`.
+
+Before invoking `Commit`, Tendermint locks the mempool and flushes the mempool connection. This ensures that
+no new messages
+will be received on the mempool connection during this processing step, providing a good opportunity to safely update all four
```

Comment:
there is only one connection, and the mempool connection is not separate from the other connections.